xbmc/dialogs/GUIDialogBusy.cpp

```diff
@@ -79,30 +76,30 @@ bool CGUIDialogBusy::WaitOnEvent(CEvent &event, unsigned int displaytime /* = 10
   bool cancelled = false;
   if (!event.Wait(std::chrono::milliseconds(displaytime)))
   {
-    // throw up the progress
-    CGUIDialogBusy* dialog = CServiceBroker::GetGUI()->GetWindowManager().GetWindow<CGUIDialogBusy>(WINDOW_DIALOG_BUSY);
+    CGUIDialogBusy* dialog = static_cast<CGUIDialogBusy*>(
+        CServiceBroker::GetGUI()->GetWindowManager().GetWindow(WINDOW_DIALOG_BUSY));
     if (dialog)
     {
-      if (dialog->IsDialogRunning())
+      if (++dialog->m_waiters == 1)
       {
-        CLog::Log(LOGFATAL, "Logic error due to two concurrent busydialogs, this is a known issue. "
-                            "The application will exit.");
-        throw std::logic_error("busy dialog already running");
+        dialog->Open();
       }
 
-      dialog->Open();
-
       while (!event.Wait(1ms))
       {
         dialog->ProcessRenderLoop(false);
-        if (allowCancel && dialog->IsCanceled())
```

Comment:
Consider a boarder case where CGUIDialogBusy::WaitOnEvent is called a thousand times from components running on the main thread. A thousand waiters will call ProcessRenderLoop and and each one will render the GUI and drive the process loop. This will obviously not work. (it won't even work correctly if there would be only 2 waiters)

Reply:
> A thousand waiters will call ProcessRenderLoop and and each one will render the GUI and drive the process loop. 

I do not see the problem here. There is only one active "ProcessRenderLoop loop" at a time, the one from the top-most waiter on the stack, all others are just waiting for their call to ProcessRenderLoop to return.

Reply:
CBusyDialog::WaitOnEvent -> dialog->ProcessRenderLoop(false); (waits for ProcessRenderLoop to return) 
-> CBusyDialog::WaitOnEvent -> dialog->ProcessRenderLoop(false); (and wait for ProcessRenderLoop to return)-> l
-> CBusyDialog::WaitOnEvent ->dialog->ProcessRenderLoop(false); (rendering in a loop until cancelled or event was set)

Reply:
> all others are just waiting for their call to ProcessRenderLoop to return.

where do you think they are waiting?

```
bool CGUIWindowManager::ProcessRenderLoop(bool renderOnly)
{
  bool renderGui = true;

  if (CServiceBroker::GetAppMessenger()->IsProcessThread() && m_pCallback)
  {
    renderGui = m_pCallback->GetRenderGUI();
    m_iNested++;
    if (!renderOnly)
      m_pCallback->Process();
    m_pCallback->FrameMove(!renderOnly);
    m_pCallback->Render();
    m_iNested--;
  }
  if (g_application.m_bStop || !renderGui)
    return false;
  else
    return true;
}
```

The top most waiter calls m_pCallback->Process(), which may call a function that calls ProcessRenderLoop again. This will end up having multiple render loops on the stack. This was the reason for making the 2nd call bail out.

Reply:
@ksooo some years ago I spent quite some time on investigating hard to debug issues. I found that having multiple render loops on the stack lead to unpredictable behaviour even cycles after the stack got cleared. Now as you set it back to the state it was before, the issues will come back as well.

Reply:
But what exactly was that unpredictable behavior? Back then when you introduced the exception I was asking for an explanation: https://github.com/xbmc/xbmc/pull/13833#issuecomment-543170237

I think I understand what you ment and think my PR fixes the „not cancallable“ problem you mentioned. 

On first Open, the window gets registered in the map of active windows at window manager
On second Open, nothing will be inserted in that map, because the key is the window id and from the first call there is already an entry.
On Close from the logical second dialog, the entry will be removed from the window mgr map
And that is the actual problem, because from now on no actions will be dispatched to the logical first dialog (as this is done by the window mgr via the map of active windows
This is what fernet means with “cannot be cancelled” - abort action comes not through to the first busy dialog
My solution is simply not to call open/close if there is already a busy dialog active, but rather to jump straight into the renderlopp-loop, logically exactly what you did to solve the problem in cmediawindow.

Reply:
The problem is not the dialog itself. The problem is the render loop. If you have more than a single instance of CGUIWindowManager::ProcessRenderLoop on the stack, the strict ordering of FrameMove, Render gets broken because one of those methods may call the BusyDialog again.